frame/mixnet/src/lib.rs

```diff
@@ -0,0 +1,675 @@
+// Misc helper funcs
+////////////////////////////////////////////////////////////////////////////////
+
+fn check_removed_all(res: MultiRemovalResults) {
+	debug_assert!(res.maybe_cursor.is_none());
+}
+
+fn twox<BlockNumber: UniqueSaturatedInto<u64>>(
+	block_number: BlockNumber,
+	kx_public: &KxPublic,
+) -> u64 {
+	let block_number: u64 = block_number.unique_saturated_into();
+	let mut data = [0; 8 + KX_PUBLIC_SIZE];
+	*array_mut_ref![data, 0, 8] = block_number.to_le_bytes();
+	*array_mut_ref![data, 8, KX_PUBLIC_SIZE] = *kx_public;
+	u64::from_le_bytes(sp_io::hashing::twox_64(&data))
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// The pallet
+////////////////////////////////////////////////////////////////////////////////
+
+/// `$session_index` must be the index of the current session. If `$current` is `true`, evaluates
+/// `$expr` with `$mixnodes<T>` aliased to the current mixnode set. Otherwise, evaluates `$expr`
+/// with `$mixnodes<T>` aliased to the other mixnode set (which is either the next or previous
+/// mixnode set, depending on the session phase).
+macro_rules! with_mixnodes {
+	($session_index:expr, $current:expr, $mixnodes:ident, $expr:expr) => {
+		if (($session_index & 1) != 0) == $current {
+			type $mixnodes<T> = OddSessionMixnodes<T>;
+			$expr
+		} else {
+			type $mixnodes<T> = EvenSessionMixnodes<T>;
+			$expr
+		}
+	};
+}
+
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + SendTransactionTypes<Call<Self>> {
+		/// The maximum number of authorities per session.
+		#[pallet::constant]
+		type MaxAuthorities: Get<AuthorityIndex>;
+
+		/// The maximum size of one of a mixnode's external addresses.
+		#[pallet::constant]
+		type MaxExternalAddressSize: Get<u32>;
+
+		/// The maximum number of external addresses for a mixnode.
+		#[pallet::constant]
+		type MaxExternalAddressesPerMixnode: Get<u32>;
+
+		/// Just for retrieving the current session index.
+		type ValidatorSet: ValidatorSet<Self::AccountId>;
+
+		/// Session progress/length estimation. Used to determine when to send registration
+		/// transactions and the longevity of these transactions.
+		type NextSessionRotation: EstimateNextSessionRotation<BlockNumberFor<Self>>;
+
+		/// Length of the `CoverToCurrent` phase of each session, in blocks.
+		#[pallet::constant]
+		type NumCoverToCurrentBlocks: Get<BlockNumberFor<Self>>;
+
+		/// Length of the `RequestsToCurrent` phase of each session, in blocks.
+		#[pallet::constant]
+		type NumRequestsToCurrentBlocks: Get<BlockNumberFor<Self>>;
+
+		/// Length of the `CoverToPrev` phase of each session, in blocks.
+		#[pallet::constant]
+		type NumCoverToPrevBlocks: Get<BlockNumberFor<Self>>;
```

Comment:
There should be some comment on the order of the phases.

Reply:
Done.